standard/types.md

```diff
@@ -705,3 +714,155 @@ An *unmanaged_type* is any type that isn’t a *reference_type*, a *type_paramet
+- A reference variable of a non-nullable type is assigned to an expression that is *maybe null*.
+- The `?` is used to note a nullable reference type.
+- The null-forgiving operator `!` (§Null-Forgiving-Expressions) is used to set the null state of its operand to *not null*.
+
+> *Example*:
+>
+> <!-- Example: {template:"code-in-main-without-using", name:"NullableAnnotationContext3", ignoredWarnings:["CS0219"], expectedWarnings:["CS8632", "CS8602"], expectedException:"NullReferenceException"} -->
+> ```csharp
+> #nullable disable annotations
+> #nullable enable warnings
+> string? s1 = null;    // OK; ? makes s2 nullable
+> string s2 = null;     // OK; null-state of s2 is "maybe null"
+> s2 = null;            // OK; null-state of s2 is "maybe null"
+> char c1 = s2[1];      // Warning; dereference of a possible null; throws NullReferenceException
+> c1 = s2![1];          // The warning is suppressed
+> ```
+>
+> *end example*
+
+#### §Nullable-enable-Context Nullable enable
+
+When both the warning flag and the annotations flag are enabled, the nullable context is *enabled*.
+
+When the nullable context is ***enabled***:
+
+- For any reference type `T`, the annotation `?` in `T?` makes `T?` a nullable type, whereas the unannotated `T` is non-nullable.
+- The compiler can use static flow analysis to determine the null state of any reference variable. When nullable warnings are enabled, a reference variable’s null state (§Nullabilities-And-Null-States) is either *not null*, *maybe null*, or *maybe default* and
+- The null-forgiving operator `!` (§Null-Forgiving-Expressions) sets the null state of its operand to *not null*.
+
+### §Nullabilities-And-Null-States Nullabilities and null states
+
+A compiler is not required to perform any static analysis nor is it required to generate any diagnostic messages related to nullability.
+
+**The remainder of this subclause is conditionally normative.**
+
+Every expression has one of three ***null state***s:
+
+- *maybe null*: The value of the expression may evaluate to null.
+- *maybe default*: The value of the expression may evaluate to the default value for that type.
+- *not null*: The value of the expression isn't null.
+
+The ***default null state*** of an expression is determined by its type, and the state of the annotations flag when it is declared:
+
+- The default null state of a nullable reference type is:
+  - Maybe null when its declaration is in text where the annotations flag is enabled.
+  - Not null when its declaration is in text where the annotations flag is disabled.
+- The default null state of a non-nullable reference type is not null.
+
+A diagnostic can be produced when a variable (§9.2.1) of a non nullable reference type is initialized or assigned to an expression that is maybe null when that variable is declared in text where the annotation flag is enabled.
+
+The compiler can update the null state of a variable when the variable is read or written to.
```

Comment:
Only then? How about this:

```csharp
string? x = null;
ref string? y = ref x;
x = "hello";
Console.WriteLine(y.Length);
```

Can the compiler (theoretically) update the null state of `y` on the assignment to `x`, if it's smart enough to work out that they refer to the same storage location?